src/ring_theory/finiteness.lean

```diff
@@ -0,0 +1,271 @@
+
+In this file we define several notions of finiteness that are common in commutative algebra.
+
+## Main declarations
+
+- `module.finite`, `algebra.finite`, `ring_hom.finite`, `alg_hom.finite`
+  all of these express that some object is finitely generated *as module* over some base ring.
+- `algebra.finite_type`, `ring_hom.finite_type`, `alg_hom.finite_type`
+  all of these express that some object is finitely generated *as algebra* over some base ring.
+
+-/
+
+open function (surjective)
+open_locale big_operators
+
+section module_and_algebra
+
+variables (R A B M N : Type*) [comm_ring R]
+variables [comm_ring A] [algebra R A] [comm_ring B] [algebra R B]
+variables [add_comm_group M] [module R M]
+variables [add_comm_group N] [module R N]
+
+/-- A module over a commutative ring is `finite` if it is finitely generated as a module. -/
+@[class]
+def module.finite : Prop := (⊤ : submodule R M).fg
+
+/-- An algebra over a commutative ring is of `finite_type` if it is finitely generated
+over the base ring as algebra. -/
+@[class]
+def algebra.finite_type : Prop := (⊤ : subalgebra R A).fg
+
+namespace module
+
+variables {R M N}
+lemma finite_def : finite R M ↔ (⊤ : submodule R M).fg := iff.rfl
+variables (R M N)
+
+@[priority 100] -- see Note [lower instance priority]
+instance is_noetherian.finite [is_noetherian R M] : finite R M :=
+is_noetherian.noetherian ⊤
+
+namespace finite
+
+variables {R M N}
+
+lemma of_surjective [hM : finite R M] (f : M →ₗ[R] N) (hf : surjective f) :
+  finite R N :=
+by { rw [finite, ← linear_map.range_eq_top.2 hf, ← submodule.map_top], exact submodule.fg_map hM }
+
+lemma of_injective [is_noetherian R N] (f : M →ₗ[R] N)
+  (hf : function.injective f) : finite R M :=
+fg_of_injective f $ linear_map.ker_eq_bot.2 hf
+
+variables (R)
+
+instance self : finite R R :=
+⟨{1}, by simpa only [finset.coe_singleton] using ideal.span_singleton_one⟩
+
+variables {R}
+
+instance prod [hM : finite R M] [hN : finite R N] : finite R (M × N) :=
+begin
+  rw [finite, ← submodule.prod_top],
+  exact submodule.fg_prod hM hN
+end
+
+lemma equiv [hM : finite R M] (e : M ≃ₗ[R] N) : finite R N :=
+of_surjective (e : M →ₗ[R] N) e.surjective
+
+section algebra
+
+lemma trans [algebra A B] [is_scalar_tower R A B] [hRA : finite R A] [hAB : finite A B] :
+  finite R B :=
+let ⟨s, hs⟩ := hRA, ⟨t, ht⟩ := hAB in submodule.fg_def.2
+⟨set.image2 (•) (↑s : set A) (↑t : set B),
+set.finite.image2 _ s.finite_to_set t.finite_to_set,
+by rw [set.image2_smul, submodule.span_smul hs (↑t : set B), ht, submodule.restrict_scalars'_top]⟩
+
+@[priority 100] -- see Note [lower instance priority]
+instance finite_type [hRA : finite R A] : algebra.finite_type R A :=
+subalgebra.fg_of_submodule_fg hRA
+
+end algebra
+
+end finite
+
+end module
+
+namespace algebra
+
+namespace finite_type
+
+lemma self : finite_type R R := ⟨{1}, subsingleton.elim _ _⟩
+
+section
+open_locale classical
+attribute [elab_as_eliminator] mv_polynomial.induction_on'
```

Comment:
Should this attribute really be set where induction_on' is defined? It's the "correct" attribute, right?

Reply:
@kckennylau :arrow_up:

Reply:
Yes, it should.

Reply:
Fixed.